### ISSUE TITLE
Query parameters objects (like params[variable]=toto, ....)

### ISSUE THREAD

**Issue by Nox-404** [OAI/OpenAPI-Specification#595]
Hi,
I'm gonna be a bit lazy and just post it back here, so here is a question/suggestion/request

Is there a way to describe object parameters like this :
- filter[name]=a
- filter[ip]=10.
- ...

What i got :
![Before](http://krondor.fr/img/actual.png)

What i need :
![Before](http://krondor.fr/img/new.png)

> The field between the brackets should be able to be a select (enum) or an input field
> 
> And when one is filled, it should add another one beneath it (and with an enum it should remove or not the selected value (configurable))

I originally posted it [there](https://github.com/swagger-api/swagger-ui/issues/2027)

**Comment by whitlockjc** [comment]
I think this is more about tooling support.  The OpenAPI specification allows you to describe parameters the way you requested already so nothing needs to change.  Since you've already created an issue in swagger-ui *(https://github.com/swagger-api/swagger-ui/issues/2027)*, that's the best course of action.

Since you were already talking to @webron in the swagger-ui issue, I'll let him decide how to handle this issue.

**Comment by webron** [comment]
If I understand what @Nox-404 is asking, that's not really supported by the spec now. Query parameter *names* are not parameterized and can't be controlled that way. You can define an enum for the values, but not for the name.

What concerns me here is the control on the type of the parameter. So if you have `name` it'd be a string, but if you have `age` it'll be an integer? Not sure how we'd represent that elegantly.

**Comment by Nox-404** [comment]
@webron yes that's what i mean, and i don't have any idea on how to describe it yet too

**Comment by webron** [comment]
Parent: #565

**Comment by whitlockjc** [comment]
I don't understand how this can't be done by OpenAPI already.  Take a look at https://github.com/apigee-127/swagger-tools/issues/294 and you'll see that people are already doing this, although it's working due to the fact that the query string parsing library supports it.  But in the referenced issue, you'll see how some are already doing this and it was a tooling issue to support it.

**Comment by webron** [comment]
That's a fair point. It could be a display issue only. Tools could have a way to collect `filter[*]` entries under one roof, though again, there's a problem with the description and the type. 

It could also be a convenience issue - instead of having to repeat the same definition for each filter with a fixed name, you can have one with an enum of 'types' for the filter. Not sure this is something we need to support, but it's fair to consider.

**Comment by raymondfeng** [comment]
We have to workaround the issue today by passing stringified json to complex query parameters, for example:
```
filter={"where":{"name":"John"}}
```

**Comment by Vanuan** [comment]
@raymondfeng How do you define such format?

I tried this in the swagger editor:

```
      parameters:
        - name: filter
          in: query
          required: false
          schema:
            $ref: '#/definitions/Person'
```

But it complains that `in: query` isn't supported for object types:

```
No enum match for: query
```

**Comment by fehguy** [comment]
@raymondfeng @Vanuan please see here:

https://github.com/OAI/OpenAPI-Specification/pull/654

For the next spec, you will be allowed complex parameters for all locations.  What's left is the `serializationStrategy`, which is quite different from `produces` and `consumes`.  If you have thoughts on defining that, it'd be helpful.

**Comment by Vanuan** [comment]
I don't quite understand what produces and consumes are used for and how are they related to serializationStrategy.

**Comment by fehguy** [comment]
See #804

**Comment by itsjavi** [comment]
For now, until we have schemas available for all parameters in t he 3.0 spec, one option is to use a string like: `?filter=with_pictures(true),ids(1,2,3),location.latitude(51.5033640)` this is URL safe and it is shorter and simpler than using brackets.

Of course the server should be able to interpret them, but for new APIs I would recommend it.

**Comment by ravanscafi** [comment]
is it possible nowadays? i'm a little bit lost
